File: src/providers/resend/rules/batch-size-not-enforced.ts
/**
 * resend-batch-size-not-enforced (reliability)
 *
 * Audit finding F [MEDIUM]: resend.batch.send accepts at most 100 emails per
 * call. Flags `batch.send(<variable>)` when the enclosing function has no
 * `<variable>.length` guard.
 *
 * To avoid false positives on chunking, calls inside a loop are skipped (the
 * loop typically slices the array into <=100-sized chunks). Calls with a
 * literal array argument are also skipped (size is statically known).
 */
import { contains, isResendBatchSendCall, startOffset } from '../utils.js';

const FUNCTION_TYPES = new Set([
  'FunctionDeclaration',
  'FunctionExpression',
  'ArrowFunctionExpression',
]);
const LOOP_TYPES = new Set([
  'ForStatement',
  'ForOfStatement',
  'ForInStatement',
  'WhileStatement',
  'DoWhileStatement',
]);
const COMPARISON_OPERATORS = new Set(['>', '>=', '<', '<=', '===', '!==', '==', '!=']);

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Enforce the 100-email batch limit before calling resend.batch.send',
      category: 'reliability',
      rationale:
        'resend.batch.send accepts at most 100 emails per call. Passing a user- or data-driven array without a length guard means the request fails outright once the list grows past 100, so an entire batch of notifications silently never sends. Guarding the array length (or chunking it into <=100-sized slices) keeps the send reliable as volume scales.',
      docsUrl: 'https://resend.com/docs/api-reference/emails/send-batch-emails',
      recommended: true,
    },
    messages: {
      batchSizeNotEnforced:
        'resend.batch.send has a 100-email limit. Guard the array length (e.g. if (items.length > 100)) before sending.',
    },
    schema: [],
  },
  create(context: any) {
    const functions: any[] = [];
    const loops: any[] = [];
    // name -> list of nodes where `<name>.length` is compared against a number.
    const lengthChecks = new Map<string, any[]>();
    const batchCalls: { node: any; argName: string }[] = [];

    function isLengthOfName(member: any): string | null {
      if (member?.type !== 'MemberExpression') return null;
      if (member.property?.type !== 'Identifier' || member.property.name !== 'length') return null;
      if (member.object?.type !== 'Identifier') return null;
      return member.object.name;
    }

    return {
      FunctionDeclaration(node: any) {
        functions.push(node);
      },
      FunctionExpression(node: any) {
        functions.push(node);
      },
      ArrowFunctionExpression(node: any) {
        functions.push(node);
      },
      ForStatement(node: any) {
        loops.push(node);
      },
      ForOfStatement(node: any) {
        loops.push(node);
      },
      ForInStatement(node: any) {
        loops.push(node);
      },
      WhileStatement(node: any) {
        loops.push(node);
      },
      DoWhileStatement(node: any) {
        loops.push(node);
      },

      BinaryExpression(node: any) {
        if (!COMPARISON_OPERATORS.has(node.operator)) return;
        // Look for `<id>.length <op> <number>` (either side).
        for (const side of [node.left, node.right]) {
          const name = isLengthOfName(side);
          if (name) {
            const list = lengthChecks.get(name) ?? [];
            list.push(node);
            lengthChecks.set(name, list);
          }
        }
      },

      CallExpression(node: any) {
        if (!isResendBatchSendCall(node)) return;
        const arg = node.arguments?.[0];
        if (arg?.type !== 'Identifier') return; // literal arrays / expressions: out of scope
        batchCalls.push({ node, argName: arg.name });
      },

      'Program:exit'() {
        for (const { node, argName } of batchCalls) {
          // Skip calls inside any loop (chunking pattern).
          if (loops.some((loop) => contains(loop, node))) continue;

          // Find the smallest enclosing function for the call.
          const enclosing = functions
            .filter((fn) => contains(fn, node))
            .sort((a, b) => startOffset(b) - startOffset(a))[0];

          // A length check on the same identifier inside the enclosing function counts.
          const checks = lengthChecks.get(argName) ?? [];
          const guarded = enclosing
            ? checks.some((chk) => contains(enclosing, chk))
            : checks.length > 0;

          if (!guarded) {
            context.report({ node, messageId: 'batchSizeNotEnforced' });
          }
        }
      },
    };
  },
};

export const resendBatchSizeNotEnforcedRule = rule;
export default rule;

File: src/providers/resend/rules/python/batch_size_not_enforced.py
"""resend-batch-size-not-enforced

Parity with the JS rule: only flag `Batch.send(<name>)` when the argument is a
variable (not a literal list) and the enclosing function has no `len(name)`
comparison. Calls inside loops are skipped (chunking pattern).
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import is_resend_batch_send

RULE_KEY = "resend-batch-size-not-enforced"


def _enclosing_function(tree: ast.AST, node: ast.AST) -> ast.AST | None:
    parents: dict[ast.AST, ast.AST] = {}
    for parent in ast.walk(tree):
        for child in ast.iter_child_nodes(parent):
            parents[child] = parent
    cur: ast.AST | None = node
    while cur is not None:
        if isinstance(cur, (ast.FunctionDef, ast.AsyncFunctionDef)):
            return cur
        cur = parents.get(cur)
    return None


def _inside_loop(tree: ast.AST, node: ast.AST) -> bool:
    parents: dict[ast.AST, ast.AST] = {}
    for parent in ast.walk(tree):
        for child in ast.iter_child_nodes(parent):
            parents[child] = parent
    cur: ast.AST | None = node
    while cur is not None:
        if isinstance(cur, (ast.For, ast.AsyncFor, ast.While)):
            return True
        cur = parents.get(cur)
    return False


def _len_of_name(node: ast.AST) -> str | None:
    """Return name if node is len(<name>), else None."""
    if not isinstance(node, ast.Call):
        return None
    func = node.func
    if not isinstance(func, ast.Name) or func.id != "len":
        return None
    if len(node.args) != 1 or not isinstance(node.args[0], ast.Name):
        return None
    return node.args[0].id


def _length_guarded_names(fn: ast.AST) -> set[str]:
    names: set[str] = set()
    for node in ast.walk(fn):
        if not isinstance(node, ast.Compare):
            continue
        # len(x) <op> <num>  or  <num> <op> len(x)
        candidates = [node.left, *[c for c in node.comparators]]
        for side in candidates:
            name = _len_of_name(side)
            if name:
                names.add(name)
    return names


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call) or not is_resend_batch_send(node):
            continue
        if not node.args:
            continue
        arg = node.args[0]
        # Literal lists / comprehensions: size known or built inline — out of scope (JS parity).
        if not isinstance(arg, ast.Name):
            continue
        if _inside_loop(tree, node):
            continue
        enclosing = _enclosing_function(tree, node)
        scope = enclosing if enclosing is not None else tree
        if arg.id in _length_guarded_names(scope):
            continue
        line = getattr(node, "lineno", 1) or 1
        col = (getattr(node, "col_offset", 0) or 0) + 1
        out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})
    return out
